File: miners/python/miner.py
"""
miner.py — external Brycoin (BRY) miner.

Runs on any machine, on any network — it only needs HTTP(S) access to
your Brycoin server. Uses real OS processes (multiprocessing), not
threads, so it gets genuine parallel CPU usage across cores — Python's
GIL would otherwise prevent that with plain threads.

Usage:
    python miner.py --server http://euger-01.bagelhosting.net:20170 \\
        --username YOUR_USERNAME --password YOUR_PASSWORD \\
        --mode solo --threads 8

    python miner.py --server http://euger-01.bagelhosting.net:20170 \\
        --username YOUR_USERNAME --password YOUR_PASSWORD \\
        --mode pool --pool 2 --threads 8

See requirements.txt for the one dependency (requests).
"""

import argparse
import os
import time
from multiprocessing import Process, Queue, Event

import requests

from bryhash import bryhash, meets_difficulty


def worker_loop(prev_hash, height, difficulty, miner_tag, thread_index, thread_count,
                 result_queue, stop_event, rate_queue):
    nonce = thread_index
    batch = 0
    BATCH_REPORT = 1500
    while not stop_event.is_set():
        candidate = f"{prev_hash}{height}{nonce}{miner_tag}"
        h = bryhash(candidate)
        batch += 1
        if meets_difficulty(h, difficulty):
            try:
                result_queue.put((nonce, h))
            except Exception:
                pass
            return
        nonce += thread_count
        if batch >= BATCH_REPORT:
            rate_queue.put(batch)
            batch = 0
    if batch:
        rate_queue.put(batch)


class BrycoinMiner:
    def __init__(self, server, username, password, mode, pool, threads):
        self.server = server.rstrip("/")
        self.username = username
        self.password = password
        self.mode = mode
        self.pool = pool
        self.threads = threads
        self.token = None
        self.session = requests.Session()

    def login(self):
        r = self.session.post(
            f"{self.server}/api/login",
            json={"username": self.username, "password": self.password},
            timeout=15,
        )
        r.raise_for_status()
        self.token = r.json()["token"]
        print(f"Logged in as {self.username}")

    def headers(self):
        return {"Authorization": f"Bearer {self.token}"}

    def fetch_job(self):
        params = {"mode": self.mode}
        if self.mode == "pool":
            params["pool"] = self.pool
        r = self.session.get(f"{self.server}/api/job", headers=self.headers(), params=params, timeout=15)
        r.raise_for_status()
        return r.json()

    def submit(self, job, nonce, hash_hex):
        payload = {
            "mode": self.mode,
            "pool": self.pool if self.mode == "pool" else None,
            "nonce": nonce,
            "hash": hash_hex,
            "height": job["height"],
            "prevHash": job["prevHash"],
        }
        r = self.session.post(f"{self.server}/api/submit", headers=self.headers(), json=payload, timeout=15)
        return r

    def heartbeat(self, hashrate):
        try:
            self.session.post(
                f"{self.server}/api/heartbeat",
                headers=self.headers(),
                json={
                    "mode": self.mode,
                    "pool": self.pool if self.mode == "pool" else None,
                    "hashrate": hashrate,
                },
                timeout=10,
            )
        except Exception:
            pass  # heartbeat is just for the dashboard's display - never worth crashing over

    def run(self):
        self.login()
        pool_note = f" pool {self.pool}" if self.mode == "pool" else ""
        print(f"Mining {self.mode}{pool_note} with {self.threads} process(es) against {self.server}")

        last_heartbeat = 0.0
        current_hashrate = 0

        while True:
            try:
                job = self.fetch_job()
            except Exception as e:
                print(f"Couldn't fetch a job ({e}), retrying in 5s...")
                time.sleep(5)
                continue

            result_queue = Queue()
            rate_queue = Queue()
            stop_event = Event()
            procs = [
                Process(
                    target=worker_loop,
                    args=(job["prevHash"], job["height"], job["difficulty"], job["minerTag"],
                          i, self.threads, result_queue, stop_event, rate_queue),
                    daemon=True,
                )
                for i in range(self.threads)
            ]
            for p in procs:
                p.start()

            total_hashes = 0
            window_start = time.time()
            last_job_check = time.time()
            found = None

            while found is None:
                try:
                    found = result_queue.get(timeout=0.5)
                except Exception:
                    found = None

                while True:
                    try:
                        total_hashes += rate_queue.get_nowait()
                    except Exception:
                        break

                now = time.time()
                if now - window_start >= 1.0:
                    current_hashrate = int(total_hashes / (now - window_start))
                    print(f"\r{current_hashrate:,} H/s   ", end="", flush=True)
                    total_hashes = 0
                    window_start = now

                if now - last_heartbeat >= 4:
                    self.heartbeat(current_hashrate)
                    last_heartbeat = now

                if found is None and now - last_job_check >= 5:
                    last_job_check = now
                    try:
                        current = self.fetch_job()
                        if current["height"] != job["height"] or current["prevHash"] != job["prevHash"]:
                            print("\nRound changed — restarting with a fresh job.")
                            break
                    except Exception:
                        pass  # network hiccup, keep mining the job we already have

            stop_event.set()
            for p in procs:
                p.terminate()
            for p in procs:
                p.join(timeout=2)

            if found is None:
                continue  # the round moved on without us - loop back and fetch a fresh job

            nonce, hash_hex = found
            print(f"\nFound a candidate for height {job['height']} — submitting...")
            try:
                r = self.submit(job, nonce, hash_hex)
                data = r.json()
                if r.status_code == 200:
                    print(f"✅ Block #{data['height']} accepted! Reward: {data.get('reward')} BRY")
                elif data.get("error") == "stale-job":
                    print("Someone else got there first — fetching a new job.")
                else:
                    print(f"Submission rejected: {data}")
            except Exception as e:
                print(f"Submit failed: {e}")


def main():
    parser = argparse.ArgumentParser(description="External Brycoin (BRY) miner")
    parser.add_argument("--server", required=True,
                         help="Brycoin server base URL, e.g. http://euger-01.bagelhosting.net:20170")
    parser.add_argument("--username", required=True)
    parser.add_argument("--password", required=True)
    parser.add_argument("--mode", choices=["solo", "pool"], default="solo")
    parser.add_argument("--pool", type=int, choices=[1, 2, 3], default=1)
    parser.add_argument("--threads", type=int, default=None,
                         help="Number of processes to mine with (default: your CPU core count)")
    args = parser.parse_args()

    threads = args.threads or os.cpu_count() or 1
    threads = max(1, min(128, threads))

    miner = BrycoinMiner(args.server, args.username, args.password, args.mode, args.pool, threads)
    try:
        miner.run()
    except KeyboardInterrupt:
        print("\nStopped.")


if __name__ == "__main__":
    main()

File: miner.js
/**
 * miner.js — external Brycoin (BRY) miner.
 *
 * Runs on any machine, on any network — it only needs HTTP(S) access to
 * your Brycoin server. Uses worker_threads (built into Node, no extra
 * dependency) for genuine parallel CPU usage across cores.
 *
 * Usage:
 *   node miner.js --server http://euger-01.bagelhosting.net:20170 \
 *     --username YOUR_USERNAME --password YOUR_PASSWORD \
 *     --mode solo --threads 8
 *
 *   node miner.js --server http://euger-01.bagelhosting.net:20170 \
 *     --username YOUR_USERNAME --password YOUR_PASSWORD \
 *     --mode pool --pool 2 --threads 8
 *
 * Requires Node.js 18+ (for built-in fetch). No npm install needed.
 */

const path = require("path");
const os = require("os");
const { Worker } = require("worker_threads");

function parseArgs() {
  const args = { mode: "solo", pool: 1, threads: null };
  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    const key = argv[i].replace(/^--/, "");
    const val = argv[i + 1];
    if (["server", "username", "password", "mode"].includes(key)) {
      args[key] = val;
      i++;
    } else if (key === "pool" || key === "threads") {
      args[key] = parseInt(val, 10);
      i++;
    }
  }
  if (!args.server || !args.username || !args.password) {
    console.error("Usage: node miner.js --server <url> --username <name> --password <pass> [--mode solo|pool] [--pool 1-3] [--threads N]");
    process.exit(1);
  }
  if (!args.threads) args.threads = os.cpus().length || 1;
  args.threads = Math.max(1, Math.min(128, args.threads));
  return args;
}

class BrycoinMiner {
  constructor({ server, username, password, mode, pool, threads }) {
    this.server = server.replace(/\/+$/, "");
    this.username = username;
    this.password = password;
    this.mode = mode;
    this.pool = pool;
    this.threads = threads;
    this.token = null;
  }

  async login() {
    const res = await fetch(`${this.server}/api/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username: this.username, password: this.password }),
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error || "Login failed");
    this.token = data.token;
    console.log(`Logged in as ${this.username}`);
  }

  authHeaders() {
    return { Authorization: `Bearer ${this.token}`, "Content-Type": "application/json" };
  }

  async fetchJob() {
    const qs = this.mode === "pool" ? `?mode=pool&pool=${this.pool}` : `?mode=solo`;
    const res = await fetch(`${this.server}/api/job${qs}`, { headers: this.authHeaders() });
    const data = await res.json();
    if (!res.ok) throw new Error(data.error || "Failed to fetch job");
    return data;
  }

  async submit(job, nonce, hash) {
    const res = await fetch(`${this.server}/api/submit`, {
      method: "POST",
      headers: this.authHeaders(),
      body: JSON.stringify({
        mode: this.mode,
        pool: this.mode === "pool" ? this.pool : null,
        nonce,
        hash,
        height: job.height,
        prevHash: job.prevHash,
      }),
    });
    const data = await res.json().catch(() => ({}));
    return { status: res.status, data };
  }

  async heartbeat(hashrate) {
    try {
      await fetch(`${this.server}/api/heartbeat`, {
        method: "POST",
        headers: this.authHeaders(),
        body: JSON.stringify({ mode: this.mode, pool: this.mode === "pool" ? this.pool : null, hashrate }),
      });
    } catch (e) {
      // heartbeat is just for the dashboard's display - never worth crashing over
    }
  }

  spawnWorkers(job) {
    const workerPath = path.join(__dirname, "mine-worker.js");
    this.workers = [];
    this.currentHashrate = 0;
    return new Promise((resolve) => {
      let settled = false;
      for (let i = 0; i < this.threads; i++) {
        const w = new Worker(workerPath, {
          workerData: {
            prevHash: job.prevHash,
            height: job.height,
            difficulty: job.difficulty,
            minerTag: job.minerTag,
            threadIndex: i,
            threadCount: this.threads,
          },
        });
        w.on("message", (msg) => {
          if (msg.type === "rate") {
            this._rateAccum = (this._rateAccum || 0) + msg.count;
          } else if (msg.type === "found" && !settled) {
            settled = true;
            resolve({ nonce: msg.nonce, hash: msg.hash });
          }
        });
        w.on("error", (err) => console.error("Worker error:", err.message));
        this.workers.push(w);
      }
    });
  }

  stopWorkers() {
    for (const w of this.workers || []) {
      w.postMessage({ type: "stop" });
      w.terminate().catch(() => {});
    }
    this.workers = [];
  }

  async run() {
    await this.login();
    const poolNote = this.mode === "pool" ? ` pool ${this.pool}` : "";
    console.log(`Mining ${this.mode}${poolNote} with ${this.threads} worker thread(s) against ${this.server}`);

    let lastHeartbeat = 0;

    while (true) {
      let job;
      try {
        job = await this.fetchJob();
      } catch (e) {
        console.error(`Couldn't fetch a job (${e.message}), retrying in 5s...`);
        await sleep(5000);
        continue;
      }

      this._rateAccum = 0;
      const resultPromise = this.spawnWorkers(job);

      let found = null;
      const rateTimer = setInterval(() => {
        const rate = this._rateAccum || 0;
        this._rateAccum = 0;
        this.currentHashrate = rate;
        process.stdout.write(`\r${rate.toLocaleString()} H/s   `);
      }, 1000);

      const heartbeatTimer = setInterval(() => {
        this.heartbeat(this.currentHashrate || 0);
      }, 4000);

      const jobCheckTimer = setInterval(async () => {
        try {
          const current = await this.fetchJob();
          if (current.height !== job.height || current.prevHash !== job.prevHash) {
            console.log("\nRound changed — restarting with a fresh job.");
            clearInterval(jobCheckTimer);
            this.stopWorkers();
          }
        } catch (e) {
          // network hiccup, keep mining the job we already have
        }
      }, 5000);

      found = await Promise.race([
        resultPromise,
        new Promise((resolve) => {
          const check = setInterval(() => {
            if ((this.workers || []).length === 0) {
              clearInterval(check);
              resolve(null); // workers were stopped because the round changed
            }
          }, 200);
        }),
      ]);

      clearInterval(rateTimer);
      clearInterval(heartbeatTimer);
      clearInterval(jobCheckTimer);
      this.stopWorkers();

      if (!found) continue; // round moved on without us

      console.log(`\nFound a candidate for height ${job.height} — submitting...`);
      try {
        const { status, data } = await this.submit(job, found.nonce, found.hash);
        if (status === 200) {
          console.log(`✅ Block #${data.height} accepted! Reward: ${data.reward} BRY`);
        } else if (data.error === "stale-job") {
          console.log("Someone else got there first — fetching a new job.");
        } else {
          console.log("Submission rejected:", data);
        }
      } catch (e) {
        console.log("Submit failed:", e.message);
      }
    }
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  const args = parseArgs();
  const miner = new BrycoinMiner(args);
  try {
    await miner.run();
  } catch (e) {
    console.error("Fatal error:", e.message);
    process.exit(1);
  }
}

process.on("SIGINT", () => {
  console.log("\nStopped.");
  process.exit(0);
});

main();
